{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating Codon Adaptation Index according (CAI) according to Sharp et al. 1987\n",
    "\n",
    "The codon adaptation index (CAI) measures the level of adaptation of a gene to a certain organism. It was introduced by Sharp 1987. CAI measures the deviation of the codons in a given protein coding gene sequence with respect to the codons used in a reference set of genes.\n",
    "\n",
    "Synthetic genes are often made for a particular organism with an optimal CAI which often seem to improve expression. See the [wikipedia](https://en.wikipedia.org/wiki/Codon_Adaptation_Index) article on CAI.\n",
    "\n",
    "[Biopython](http://biopython.org) has a module called [CodonUsage](http://biopython.org/DIST/docs/api/Bio.SeqUtils.CodonUsage.CodonAdaptationIndex-class.html). This module does not contain any of the necessary reference data in order to calculate CAI for yeast. Worse, the fidelity of the Biopython module had been questioned in this exchange on [Biostars](https://www.biostars.org/p/290485/).\n",
    "\n",
    "The python module [CAI](https://pypi.org/project/CAI/) is an alternative implementation. \n",
    "\n",
    "The objective of this notebook is to test the fidelity of this python module using the original data of \n",
    "Sharp 1987 for yeast.\n",
    "\n",
    "Codon adaptation data is tabulated in table 1 of the [publication](https://www.ncbi.nlm.nih.gov/pubmed/3547335).\n",
    "\n",
    "![Sharp Table 1](sharp_table_1.png)\n",
    "\n",
    "Some examples of CAI numbers for identifiable yeast genes are given in table 2 of the same publication.\n",
    "\n",
    "[Sharp Table 2](sharp_table_2.png)\n",
    "\n",
    "<img src=\"sharp_table_2.png\" alt=\"Drawing\" style=\"width=200\"/>\n",
    "\n",
    "The [CAI](https://pypi.org/project/CAI/) module has a function [CAI](https://cai.readthedocs.io/en/latest/api.html#CAI.CAI) that can take a series of genes as reference or a dictionary with the triplets as keys and relative synonymous codon usage (RSCU) as values. This data is present in the Sharp 1987 table 1.\n",
    "\n",
    "Table 2 lists CAI for yeast genes GAL4, PPR1 and GPD1.\n",
    "\n",
    "The strategy will be to turn the Table 1 into a dict, use this to calculate CAI for [GAL4](https://www.yeastgenome.org/locus/S000006169), [PPR1](https://www.yeastgenome.org/locus/S000004004) and [GPD1](https://www.yeastgenome.org/locus/S000003424). The gene referred to as GPD1 in Sharp 1987 table 2 is most likely the TDH3 gene encoding Glyceraldehyde-3-phosphate dehydrogenase (GAPDH) .\n",
    "\n",
    "### References\n",
    "\n",
    "Sharp PM, Li WH. The codon Adaptation Index--a measure of directional synonymous codon usage bias, and its potential applications. Nucleic Acids Res 1987;15:1281–95.\n",
    "\n",
    "[\n",
    "Benjamin Lee. (2017). Python Implementation of Codon Adaptation Index.](http://joss.theoj.org/papers/8adf6bd9fd6391d5343d15ea0b6b6525)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import all dependencies as per these [tips](https://hackernoon.com/10-tips-on-using-jupyter-notebook-abc0ba7028a4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from CAI import CAI\n",
    "\n",
    "from pygenome import sg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Sharp table 1 was copied and edited from the pdf version of the publication. The order of the columns remains the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sharp_table_1 = \"\"\"\n",
    "AA Tri RSCUe we RSCUy wy\n",
    "Phe TTT 0.456 0.296 0.203 0.113\n",
    "Phe TTC 1.544 1.000 1.797 1.000\n",
    "Leu TTA 0.106 0.020 0.601 0.117\n",
    "Leu TTG 0.106 0.020 5.141 1.000\n",
    "Leu CTT 0.225 0.042 0.029 0.006\n",
    "Leu CTC 0.198 0.037 0.014 0.003\n",
    "Leu CTA 0.040 0.007 0.200 0.039\n",
    "Leu CTG 5.326 1.000 0.014 0.003\n",
    "Ile ATT 0.466 0.185 1.352 0.823\n",
    "Ile ATC 2.525 1.000 1.643 1.000\n",
    "Ile ATA 0.008 0.003 0.005 0.003\n",
    "Met ATG 1.000 1.000 1.000 1.000\n",
    "Val GTT 2.244 1.000 2.161 1.000\n",
    "Val GTC 0.148 0.066 1.796 0.831\n",
    "Val GTA 1.111 0.495 0.004 0.002\n",
    "Val GTG 0.496 0.221 0.039 0.018\n",
    "Ser TCT 2.571 1.000 3.359 1.000\n",
    "Ser TCC 1.912 0.744 2.327 0.693\n",
    "Ser TCA 0.198 0.077 0.122 0.036\n",
    "Ser TCG 0.044 0.017 0.017 0.005\n",
    "Pro CCT 0.231 0.070 0.179 0.047\n",
    "Pro CCC 0.038 0.012 0.036 0.009\n",
    "Pro CCA 0.442 0.135 3.776 1.000\n",
    "Pro CCG 3.288 1.000 0.009 0.002\n",
    "Thr ACT 1.804 0.965 1.899 0.921\n",
    "Thr ACC 1.870 1.000 2.063 1.000\n",
    "Thr ACA 0.141 0.076 0.025 0.012\n",
    "Thr ACG 0.185 0.099 0.013 0.006\n",
    "Ala GCT 1.877 1.000 3.005 1.000\n",
    "Ala GCC 0.228 0.122 0.948 0.316\n",
    "Ala GCA 1.099 0.586 0.044 0.015\n",
    "Ala GCG 0.796 0.424 0.004 0.001\n",
    "Tyr TAT 0.386 0.239 0.132 0.071\n",
    "Tyr TAC 1.614 1.000 1.868 1.000\n",
    "His CAT 0.451 0.291 0.394 0.245\n",
    "His CAC 1.549 1.000 1.606 1.000\n",
    "Gln CAA 0.220 0.124 1.987 1.000\n",
    "Gln CAG 1.780 1.000 0.013 0.007\n",
    "Asn AAT 0.097 0.051 0.100 0.053\n",
    "Asn AAC 1.903 1.000 1.900 1.000\n",
    "Lys AAA 1.596 1.000 0.237 0.135\n",
    "Lys AAG 0.404 0.253 1.763 1.000\n",
    "Asp GAT 0.605 0.434 0.713 0.554\n",
    "Asp GAC 1.395 1.000 1.287 1.000\n",
    "Glu GAA 1.589 1.000 1.968 1.000\n",
    "Glu GAG 0.411 0.259 0.032 0.016\n",
    "Cys TGT 0.667 0.500 1.857 1.000\n",
    "Cys TGC 1.333 1.000 0.143 0.077\n",
    "Trp TGG 1.000 1.000 1.000 1.000\n",
    "Arg CGT 4.380 1.000 0.718 0.137\n",
    "Arg CGC 1.561 0.356 0.008 0.002\n",
    "Arg CGA 0.017 0.004 0.008 0.002\n",
    "Arg CGG 0.017 0.004 0.008 0.002\n",
    "Ser AGT 0.220 0.085 0.070 0.021\n",
    "Ser AGC 1.055 0.410 0.105 0.031\n",
    "Arg AGA 0.017 0.004 5.241 1.000\n",
    "Arg AGG 0.008 0.002 0.017 0.003\n",
    "Gly GGT 2.283 1.000 3.898 1.000\n",
    "Gly GGC 1.652 0.724 0.077 0.020\n",
    "Gly GGA 0.022 0.010 0.009 0.002\n",
    "Gly GGG 0.043 0.019 0.017 0.004\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas can turn a text table into a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sharp_df = pd.read_csv(StringIO(sharp_table_1.strip()), sep=\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting dataframe seem correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA</th>\n",
       "      <th>Tri</th>\n",
       "      <th>RSCUe</th>\n",
       "      <th>we</th>\n",
       "      <th>RSCUy</th>\n",
       "      <th>wy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Phe</td>\n",
       "      <td>TTT</td>\n",
       "      <td>0.456</td>\n",
       "      <td>0.296</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Phe</td>\n",
       "      <td>TTC</td>\n",
       "      <td>1.544</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.797</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Leu</td>\n",
       "      <td>TTA</td>\n",
       "      <td>0.106</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.601</td>\n",
       "      <td>0.117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Leu</td>\n",
       "      <td>TTG</td>\n",
       "      <td>0.106</td>\n",
       "      <td>0.020</td>\n",
       "      <td>5.141</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Leu</td>\n",
       "      <td>CTT</td>\n",
       "      <td>0.225</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Leu</td>\n",
       "      <td>CTC</td>\n",
       "      <td>0.198</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Leu</td>\n",
       "      <td>CTA</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Leu</td>\n",
       "      <td>CTG</td>\n",
       "      <td>5.326</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Ile</td>\n",
       "      <td>ATT</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.185</td>\n",
       "      <td>1.352</td>\n",
       "      <td>0.823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ile</td>\n",
       "      <td>ATC</td>\n",
       "      <td>2.525</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.643</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Ile</td>\n",
       "      <td>ATA</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Met</td>\n",
       "      <td>ATG</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Val</td>\n",
       "      <td>GTT</td>\n",
       "      <td>2.244</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2.161</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Val</td>\n",
       "      <td>GTC</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.066</td>\n",
       "      <td>1.796</td>\n",
       "      <td>0.831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Val</td>\n",
       "      <td>GTA</td>\n",
       "      <td>1.111</td>\n",
       "      <td>0.495</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Val</td>\n",
       "      <td>GTG</td>\n",
       "      <td>0.496</td>\n",
       "      <td>0.221</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ser</td>\n",
       "      <td>TCT</td>\n",
       "      <td>2.571</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3.359</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ser</td>\n",
       "      <td>TCC</td>\n",
       "      <td>1.912</td>\n",
       "      <td>0.744</td>\n",
       "      <td>2.327</td>\n",
       "      <td>0.693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Ser</td>\n",
       "      <td>TCA</td>\n",
       "      <td>0.198</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.122</td>\n",
       "      <td>0.036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ser</td>\n",
       "      <td>TCG</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pro</td>\n",
       "      <td>CCT</td>\n",
       "      <td>0.231</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pro</td>\n",
       "      <td>CCC</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pro</td>\n",
       "      <td>CCA</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.135</td>\n",
       "      <td>3.776</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Pro</td>\n",
       "      <td>CCG</td>\n",
       "      <td>3.288</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Thr</td>\n",
       "      <td>ACT</td>\n",
       "      <td>1.804</td>\n",
       "      <td>0.965</td>\n",
       "      <td>1.899</td>\n",
       "      <td>0.921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Thr</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1.870</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2.063</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Thr</td>\n",
       "      <td>ACA</td>\n",
       "      <td>0.141</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Thr</td>\n",
       "      <td>ACG</td>\n",
       "      <td>0.185</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Ala</td>\n",
       "      <td>GCT</td>\n",
       "      <td>1.877</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3.005</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Ala</td>\n",
       "      <td>GCC</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.122</td>\n",
       "      <td>0.948</td>\n",
       "      <td>0.316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Ala</td>\n",
       "      <td>GCG</td>\n",
       "      <td>0.796</td>\n",
       "      <td>0.424</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Tyr</td>\n",
       "      <td>TAT</td>\n",
       "      <td>0.386</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.132</td>\n",
       "      <td>0.071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Tyr</td>\n",
       "      <td>TAC</td>\n",
       "      <td>1.614</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.868</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>His</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0.451</td>\n",
       "      <td>0.291</td>\n",
       "      <td>0.394</td>\n",
       "      <td>0.245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>His</td>\n",
       "      <td>CAC</td>\n",
       "      <td>1.549</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.606</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Gln</td>\n",
       "      <td>CAA</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.124</td>\n",
       "      <td>1.987</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Gln</td>\n",
       "      <td>CAG</td>\n",
       "      <td>1.780</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Asn</td>\n",
       "      <td>AAT</td>\n",
       "      <td>0.097</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Asn</td>\n",
       "      <td>AAC</td>\n",
       "      <td>1.903</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.900</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Lys</td>\n",
       "      <td>AAA</td>\n",
       "      <td>1.596</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.237</td>\n",
       "      <td>0.135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Lys</td>\n",
       "      <td>AAG</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.253</td>\n",
       "      <td>1.763</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Asp</td>\n",
       "      <td>GAT</td>\n",
       "      <td>0.605</td>\n",
       "      <td>0.434</td>\n",
       "      <td>0.713</td>\n",
       "      <td>0.554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Asp</td>\n",
       "      <td>GAC</td>\n",
       "      <td>1.395</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.287</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Glu</td>\n",
       "      <td>GAA</td>\n",
       "      <td>1.589</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.968</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Glu</td>\n",
       "      <td>GAG</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.259</td>\n",
       "      <td>0.032</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Cys</td>\n",
       "      <td>TGT</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.500</td>\n",
       "      <td>1.857</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Cys</td>\n",
       "      <td>TGC</td>\n",
       "      <td>1.333</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Trp</td>\n",
       "      <td>TGG</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Arg</td>\n",
       "      <td>CGT</td>\n",
       "      <td>4.380</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Arg</td>\n",
       "      <td>CGC</td>\n",
       "      <td>1.561</td>\n",
       "      <td>0.356</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Arg</td>\n",
       "      <td>CGA</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Arg</td>\n",
       "      <td>CGG</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Ser</td>\n",
       "      <td>AGT</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Ser</td>\n",
       "      <td>AGC</td>\n",
       "      <td>1.055</td>\n",
       "      <td>0.410</td>\n",
       "      <td>0.105</td>\n",
       "      <td>0.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Arg</td>\n",
       "      <td>AGA</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.004</td>\n",
       "      <td>5.241</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Arg</td>\n",
       "      <td>AGG</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Gly</td>\n",
       "      <td>GGT</td>\n",
       "      <td>2.283</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3.898</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Gly</td>\n",
       "      <td>GGC</td>\n",
       "      <td>1.652</td>\n",
       "      <td>0.724</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Gly</td>\n",
       "      <td>GGA</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Gly</td>\n",
       "      <td>GGG</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     AA  Tri  RSCUe     we  RSCUy     wy\n",
       "0   Phe  TTT  0.456  0.296  0.203  0.113\n",
       "1   Phe  TTC  1.544  1.000  1.797  1.000\n",
       "2   Leu  TTA  0.106  0.020  0.601  0.117\n",
       "3   Leu  TTG  0.106  0.020  5.141  1.000\n",
       "4   Leu  CTT  0.225  0.042  0.029  0.006\n",
       "5   Leu  CTC  0.198  0.037  0.014  0.003\n",
       "6   Leu  CTA  0.040  0.007  0.200  0.039\n",
       "7   Leu  CTG  5.326  1.000  0.014  0.003\n",
       "8   Ile  ATT  0.466  0.185  1.352  0.823\n",
       "9   Ile  ATC  2.525  1.000  1.643  1.000\n",
       "10  Ile  ATA  0.008  0.003  0.005  0.003\n",
       "11  Met  ATG  1.000  1.000  1.000  1.000\n",
       "12  Val  GTT  2.244  1.000  2.161  1.000\n",
       "13  Val  GTC  0.148  0.066  1.796  0.831\n",
       "14  Val  GTA  1.111  0.495  0.004  0.002\n",
       "15  Val  GTG  0.496  0.221  0.039  0.018\n",
       "16  Ser  TCT  2.571  1.000  3.359  1.000\n",
       "17  Ser  TCC  1.912  0.744  2.327  0.693\n",
       "18  Ser  TCA  0.198  0.077  0.122  0.036\n",
       "19  Ser  TCG  0.044  0.017  0.017  0.005\n",
       "20  Pro  CCT  0.231  0.070  0.179  0.047\n",
       "21  Pro  CCC  0.038  0.012  0.036  0.009\n",
       "22  Pro  CCA  0.442  0.135  3.776  1.000\n",
       "23  Pro  CCG  3.288  1.000  0.009  0.002\n",
       "24  Thr  ACT  1.804  0.965  1.899  0.921\n",
       "25  Thr  ACC  1.870  1.000  2.063  1.000\n",
       "26  Thr  ACA  0.141  0.076  0.025  0.012\n",
       "27  Thr  ACG  0.185  0.099  0.013  0.006\n",
       "28  Ala  GCT  1.877  1.000  3.005  1.000\n",
       "29  Ala  GCC  0.228  0.122  0.948  0.316\n",
       "..  ...  ...    ...    ...    ...    ...\n",
       "31  Ala  GCG  0.796  0.424  0.004  0.001\n",
       "32  Tyr  TAT  0.386  0.239  0.132  0.071\n",
       "33  Tyr  TAC  1.614  1.000  1.868  1.000\n",
       "34  His  CAT  0.451  0.291  0.394  0.245\n",
       "35  His  CAC  1.549  1.000  1.606  1.000\n",
       "36  Gln  CAA  0.220  0.124  1.987  1.000\n",
       "37  Gln  CAG  1.780  1.000  0.013  0.007\n",
       "38  Asn  AAT  0.097  0.051  0.100  0.053\n",
       "39  Asn  AAC  1.903  1.000  1.900  1.000\n",
       "40  Lys  AAA  1.596  1.000  0.237  0.135\n",
       "41  Lys  AAG  0.404  0.253  1.763  1.000\n",
       "42  Asp  GAT  0.605  0.434  0.713  0.554\n",
       "43  Asp  GAC  1.395  1.000  1.287  1.000\n",
       "44  Glu  GAA  1.589  1.000  1.968  1.000\n",
       "45  Glu  GAG  0.411  0.259  0.032  0.016\n",
       "46  Cys  TGT  0.667  0.500  1.857  1.000\n",
       "47  Cys  TGC  1.333  1.000  0.143  0.077\n",
       "48  Trp  TGG  1.000  1.000  1.000  1.000\n",
       "49  Arg  CGT  4.380  1.000  0.718  0.137\n",
       "50  Arg  CGC  1.561  0.356  0.008  0.002\n",
       "51  Arg  CGA  0.017  0.004  0.008  0.002\n",
       "52  Arg  CGG  0.017  0.004  0.008  0.002\n",
       "53  Ser  AGT  0.220  0.085  0.070  0.021\n",
       "54  Ser  AGC  1.055  0.410  0.105  0.031\n",
       "55  Arg  AGA  0.017  0.004  5.241  1.000\n",
       "56  Arg  AGG  0.008  0.002  0.017  0.003\n",
       "57  Gly  GGT  2.283  1.000  3.898  1.000\n",
       "58  Gly  GGC  1.652  0.724  0.077  0.020\n",
       "59  Gly  GGA  0.022  0.010  0.009  0.002\n",
       "60  Gly  GGG  0.043  0.019  0.017  0.004\n",
       "\n",
       "[61 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sharp_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A dict is made from the first and fifth columns. The fifth column contain the yeast data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "RSCU_sharp = dict(zip(sharp_df[\"Tri\"],sharp_df[\"RSCUy\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "pickle.dump(RSCU_sharp,open(\"RSCU_sharp.pickle\",\"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "GAL4str = str(sg.stdgene[\"GAL4\"].cds.seq)\n",
    "PPR1str = str(sg.stdgene[\"PPR1\"].cds.seq)\n",
    "GPD1str = str(sg.stdgene[\"TDH3\"].cds.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "| Gene | CAI   |\n",
    "|------|-------|\n",
    "| GAL4 | 0.116 |\n",
    "| PPR1 | 0.114 |\n",
    "| GPD1 | 0.929 |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.116\n",
      "0.115\n",
      "0.924\n"
     ]
    }
   ],
   "source": [
    "print(round(CAI(GAL4str,RSCUs=RSCU_sharp),3))\n",
    "print(round(CAI(PPR1str,RSCUs=RSCU_sharp),3)) \n",
    "print(round(CAI(GPD1str,RSCUs=RSCU_sharp),3)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are small differences for PPR1 and GPD1, these could perhaps be explained by rounding of the RSCU values in table 1. The genes used to create the RSCU data in Sharp 1987 were described in [Sharp 1986](https://www.ncbi.nlm.nih.gov/pubmed/3526280)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![](table1a.png)\n",
    "\n",
    "![](table1b.png)\n",
    "\n",
    "![](table1b2.png)\n",
    "\n",
    "![](table1c.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    Yeast - 16 ribosomal protein genes, TEF 1, 2 enolase\n",
    "    genes, 2 GA-3-PDH genes, ADH 1, PGK, pyruvate kinase (data\n",
    "    sources given in Ref.5)\n",
    "\n",
    "\n",
    "    Ribosomal protein L16 175 0.83 0.70 0.80 0.79 (1) *1   RPL10\n",
    "    Ribosomal protein L17a 138 0.79 0.68 0.72 0.63 (2)     RPL17A\n",
    "    Ribosomal protein L25 138 0.86 0.72 0.82 0.52 (3)      RPL25\n",
    "    Ribosomal protein L29 150 0.79 0.73 0.83 0.66 (4)      RPL29\n",
    "    Ribosomal protein L34 114 0.84 0.75 0.79 0.57 (5)      RPL31A\n",
    "    Ribosomal protein 13 388 0.89 0.78 0.86 0.70 (7)       RPL16A\n",
    "    Ribosomal protein 28 187 0.89 0.86 0.89 0.61 (8)       RPS23A\n",
    "    Ribosomal protein 51a 137 0.87 0.85 0.86 0.68 (9)      RPS17A\n",
    "    Ribosomal protein 59 138 0.88 0.85 0.79 0.63 (10) *1   RPS14A\n",
    "    Ribosomal protein S10 238 0.94 0.86 0.86 0.67 (11)     RPS20\n",
    "    Ribosomal protein S16a 145 0.88 0.80 0.78 0.78 (12)    RPS16A\n",
    "    Ribosomal protein S24 131 0.86 0.77 0.67 0.81 (13) *4  RPS22A\n",
    "    Ribosomal protein 29 156 0.83 0.72 0.83 0.55 (86) *4   RPL29\n",
    "    Ribosomal protein 51B 137 0.83 0.78 0.83 0.68 (102) *4 RPS17B\n",
    "    Ribosomal protein L46 52 0.93 0.99 1.00 0.13 (6) *3    RPL39\n",
    "    Ribosomal protein S33 68 0.63 0.68 0.79 0.31 (14)      RPS28A\n",
    "    TEF 1 Elong. factor la 459 0.93 0.78 0.83 0.73 (69)    TEF1\n",
    "    enolase A 438 0.93 0.82 0.85 0.78 (29)                 ENO1\n",
    "    enolase B 438 0.96 0.85 0.86 0.75 (30)                 ENO2\n",
    "    GA-3-PDH 1 331 0.99 0.86 0.86 0.81 (34)                TDH3\n",
    "    GA-3-PDH 3 331 0.94 0.75 0.81 0.78 (35)                TDH1\n",
    "    ADH 1 349 0.91 0.76 0.79 0.74 (16)                     ADH1\n",
    "    PGK 417 0.91 0.75 0.85 0.75 (67)                       PGK1\n",
    "    Pyruvate kinase 500 0.95 0.79 0.87 0.77 (53) *4        CDC19\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "highly_expressed_genes= '''\n",
    "RPL10\n",
    "RPL17A\n",
    "RPL25\n",
    "RPL29 \n",
    "RPL31A\n",
    "RPL16A\n",
    "RPL29 \n",
    "RPS17B\n",
    "RPL39 \n",
    "RPS28A\n",
    "RPS23A\n",
    "RPS17A\n",
    "RPS14A\n",
    "RPS20 \n",
    "RPS16A\n",
    "RPS22A\n",
    "\n",
    "TEF1  \n",
    "ENO1 \n",
    "ENO2 \n",
    "TDH3 \n",
    "TDH2 \n",
    "ADH1 \n",
    "PGK1 \n",
    "CDC19\n",
    "'''\n",
    "reference = []\n",
    "for g in highly_expressed_genes.split():\n",
    "    reference.append(str(sg.stdgene[g].cds.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.104\n",
      "0.109\n",
      "0.913\n"
     ]
    }
   ],
   "source": [
    "print(round(CAI(GAL4str,reference=reference),3))\n",
    "print(round(CAI(PPR1str,reference=reference),3)) \n",
    "print(round(CAI(GPD1str,reference=reference),3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the raw data supposedly used in Sharp 1987 has a worse fit than using the table data.\n",
    "This could perhaps be due to misidentification of some genes in Sharp 1986. Gene names have changed quite a bit \n",
    "since then. Going through the references in the Sharp 1986 tables could perhaps help solving this. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bjorn38]",
   "language": "python",
   "name": "conda-env-bjorn38-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
